### PR TITLE
gRPC: use gzip level-2 compression by default

### PR DESCRIPTION
**Pull request by zkkxu**
### What is changed and how it works?
Issue Number:  Ref #12929

this pr only sacrifice 5% CPU to cut off 50% cross az network traffic between tikvs

What's Changed:
```commit-message
add initial arguments for gzip compression:
gzip_compression_level: represent gzip compression level, the origin gzip compression level is 6 and hardcoding;
compression_lower_bound: this represent gzip will compress the data only larger than this
```

### Check List 

Tests 
- No code

Side effects
- Performance regression
    - Only cost more 5% CPU

### Release note 

```release-note
none
```

### CODE REVIEW

File: src/server/config.rs [hunk 1]
```diff
@@ -26,6 +26,8 @@ const DEFAULT_GRPC_CONCURRENT_STREAM: i32 = 1024;
 const DEFAULT_GRPC_RAFT_CONN_NUM: usize = 1;
 const DEFAULT_GRPC_MEMORY_POOL_QUOTA: u64 = isize::MAX as u64;
 const DEFAULT_GRPC_STREAM_INITIAL_WINDOW_SIZE: u64 = 2 * 1024 * 1024;
+const DEFAULT_GRPC_GZIP_COMPRESSION_LEVEL: usize = 6;
```
Comment (kevin-xianliu) [inline src/server/config.rs:29]: Should us set the default level 2 instead of 6?

File: src/server/config.rs [hunk 2]
```diff
@@ -213,6 +219,8 @@ impl Default for Config {
             raft_client_queue_size: 8192,
             raft_msg_max_batch_size: 128,
             grpc_compression_type: GrpcCompressionType::None,
+            grpc_gzip_compression_level: 6,
```
Comment (kevin-xianliu) [inline src/server/config.rs:222]: Better to use  DEFAULT_GRPC_GZIP_COMPRESSION_LEVEL and DEFAULT_GRPC_COMPRESSION_LOWER_BOUND.

File: src/server/config.rs (hunk 1 above)
Comment (zhangjinpeng1987) [reply src/server/config.rs:29]: agree

Comment (knightXun) [reply src/server/config.rs]: nice hit

**Review by BusyJay — COMMENTED** [review]
Do connections between tikv and tidb need compression?

**Comment by zkkxu** [comment]
> Do connections between tikv and tidb need compression?

that depends on tidb client

**Comment by BusyJay** [comment]
If client enables gzip, what level will tikv use?

**Comment by zkkxu** [comment]
> If client enables gzip, what level will tikv use?

tikv only decompress the data. as for the returned data, it depends on tikv grpc client

**Comment by BusyJay** [comment]
> as for the returned data, it depends on tikv grpc client

If client accepts gzip, what level will be used?

**Comment by zkkxu** [comment]
> > as for the returned data, it depends on tikv grpc client
> 
> If client accepts gzip, what level will be used?

If size less than grpc_min_message_size_to_compress, will use grpc_gzip_compression_level, default 2.

**Comment by BusyJay** [comment]
> If size larger than grpc_min_message_size_to_compress, will use grpc_gzip_compression_level, default 2. otherwise no compression

How does server know it should use level 2? I don't see it's configured anywhere.

**Comment by zkkxu** [comment]
> > If size larger than grpc_min_message_size_to_compress, will use grpc_gzip_compression_level, default 2. otherwise no compression
> 
> How does server know it should use level 2? I don't see it's configured anywhere.
grpc message has a metadata, the level will be writed into the metadata

**Comment by BusyJay** [comment]
> U can image the grpc message has a metadata, the level will be writed into the metadata

I fully understand how it work. But I don't see it's configured in this PR. There is no way for grpc know the level should be 2 instead of 3, 4, 5, 6 when responding to client.

**Comment by zkkxu** [comment]
> > U can image the grpc message has a metadata, the level will be writed into the metadata
> 
> I fully understand how it work. But I don't see it's configured in this PR. There is no way for grpc know the level should be 2 instead of 3, 4, 5, 6 when responding to client.

oh, there's no compression between tikv and tidb. this pr only compress data between tikvs

**Review by BusyJay — APPROVED** [review]

**Review by zhangjinpeng1987 — APPROVED** [review]
LGTM

**Pull request merged**